File: killer_sudoku/solver/engine/hint.py
"""HintResult: rich output from a solver rule in coach/hint mode.

A HintResult is produced by rule.as_hints() when the rule fires in hint-only
mode.  It is buffered in engine.pending_hints and read by the coaching API.
"""

from __future__ import annotations

import dataclasses

from killer_sudoku.solver.engine.types import Cell, Elimination


@dataclasses.dataclass(frozen=True)
class HintResult:
    """Rich hint produced by a single rule application instance.

    Attributes:
        rule_name:               Internal rule identifier (e.g. "MustContainOutie").
        display_name:            Short human-readable name shown in the hint list.
        explanation:             Full plain-English explanation of why the rule fires,
                                 with cell coordinates in rNcM (1-based) notation.
        highlight_cells:         Every cell involved in the reasoning — used for
                                 canvas highlighting.  Cage cells, the external
                                 qualifying cell, and the outie are all included.
        eliminations:            The candidate removals this rule would make.
        placement:               (row, col, digit) if this hint is a placement hint
                                 rather than an elimination hint.  Placement hints
                                 instruct the user to enter a digit in a cell rather
                                 than remove a candidate.  When set, eliminations is
                                 typically empty.
        virtual_cage_suggestion: (cells, total) if this hint is a T3 virtual cage
                                 suggestion.  The hint recommends the user add this
                                 derived sum constraint via POST /virtual-cages.
                                 When set, eliminations is empty and placement is None.
    """

    rule_name: str
    display_name: str
    explanation: str
    highlight_cells: frozenset[Cell]
    eliminations: list[Elimination]
    placement: tuple[int, int, int] | None = None
    virtual_cage_suggestion: tuple[frozenset[Cell], int] | None = None

    @property
    def elimination_count(self) -> int:
        """Number of candidate removals this hint would produce."""
        return len(self.eliminations)

File: web/src/engine/hint.ts
/**
 * HintResult: rich output from a solver rule in coach/hint mode.
 *
 */

import type { Cell, Elimination } from './types.js';

/** The two colours used for chain-colouring visualisation (user tool and hints). */
export type CellColour = 'blue' | 'green';

/**
 * A structurally significant cell in a chain-style hint (XY-Wing, W-Wing,
 * Two-String Kite, Skyscraper, Simple Colouring), tagged with the digit(s)
 * that matter for *that* cell and an optional wash colour. Lets different
 * chain cells carry different digits, unlike a single rule-wide
 * `patternDigits` list.
 */
export interface ChainCell {
  readonly cell: Cell;
  readonly digits: readonly number[];
  readonly colour?: CellColour;
}

/** Rich hint produced by a single rule application instance. */
export interface HintResult {
  readonly ruleName: string;
  readonly displayName: string;
  readonly explanation: string;
  /**
   * Pattern cells — rendered orange on the canvas. Cells that are also in
   * `eliminations` will be overwritten yellow. For colouring rules, chain cells
   * live in `chainCells` and should not appear here.
   */
  readonly highlightCells: readonly Cell[];
  readonly eliminations: readonly Elimination[];
  /** [row, col, digit] if this hint is a placement hint. */
  readonly placement: readonly [number, number, number] | null;
  /** [cells, total] if this hint is a virtual cage suggestion. */
  readonly virtualCageSuggestion: readonly [readonly Cell[], number] | null;
  /**
   * Per-cell digit/colour tags for chain-style rules. When a cell in
   * `highlightCells` has a matching entry here, the renderer marks that
   * cell's own `digits` instead of falling back to the rule-wide
   * `patternDigits` list, and washes it with the entry's `colour` (if set).
   * Absent for non-chain rules.
   */
  readonly chainCells?: readonly ChainCell[];
  /**
   * Cells rendered with a pale-blue wash to give unit context for the deduction
   * (e.g. "this is the unit in which the digit/tuple is unique"). Distinct from
   * `highlightCells` (orange/yellow) and `chainCells` (chain-colouring blue/green).
   */
  readonly secondaryHighlightCells?: readonly Cell[];
  /**
   * Digits key to the rule's reasoning — marked with squares in `highlightCells`.
   * Absent for most rules; the renderer then derives them from `eliminations`
   * (or `placement[2]` for placement hints). Set explicitly only when pattern
   * digits differ from elimination digits (Hidden Single/Pair/Triple/Quad).
   */
  readonly patternDigits?: readonly number[];
}

export function eliminationCount(h: HintResult): number {
  return h.eliminations.length;
}

/** Looks up the `ChainCell` entry for `cell` in `hint.chainCells`, or null if absent. */
export function findChainCell(hint: HintResult, cell: Cell): ChainCell | null {
  if (!hint.chainCells) return null;
  const [r, c] = cell;
  return hint.chainCells.find(cc => cc.cell[0] === r && cc.cell[1] === c) ?? null;
}
